### PR TITLE
Week 1 - Sidebar

### PR DESCRIPTION
**Pull request by huyenttphan**
I focused on how to load data into the HTML templates. I wrote two functions that are very similar (getMessageSubject and getMessageSender) but not sure how to combine them into one.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -3,22 +3,39 @@
...
+        <div class="sender-details">
+          <p class="email">${formatSender(sender)}</p>
+          <span class="time">${timestamp}</span>
+        </div>
+        <p class="email-subject">${subject}</p>
+        <p class="snippet">${snippet}</p>
+      </button>
+      </li>`;
+  }).join('');
+}
```
Comment (dkharlan) [inline src/index.js:31]: I like the data flow here -- simple and unidirectional.  I also like how you're mapping from thread IDs to rendered `<li>`s.  You may need to break this out as your app gets more complex, but for now this an extremely clean way to do this.

File: src/store/index.js [hunk 2]
```diff
@@ -22,4 +22,18 @@ export function getMessageSubject(message: Message): ?string {
   return subjectHeader && subjectHeader.value;
 }
 
+export function getMessageSender(message: Message): ?string {
+  const { headers } = message.payload;
+  const senderHeader = headers.find(header => header.name === 'From');
+  return senderHeader && senderHeader.value;
+}
```
Comment (dkharlan) [inline src/store/index.js:29]: Not sure if I should admit this but I prefer your approach here versus [what I did](https://github.com/dkharlan/email-client-practice-app/blob/4e54773090ffd060c6012f6ee2b1ceb7cd07b2fd/src/ui/transform.js#L6) :joy:

File: src/store/index.js [hunk 3]
```diff
@@ -22,4 +22,18 @@ export function getMessageSubject(message: Message): ?string {
...
+
+export function formatSender(sender: String){
+  return sender.split('\\')[0];
+}
+
+export function formatTimestamp(date: string){
+  return new Date(Number(date)).toLocaleDateString();
```
Comment (dkharlan) [inline src/store/index.js:36]: Locale-specific and no dependencies :+1:

**Review by dkharlan — APPROVED** [review]
The only thing I can really think of here that I'd change is that I think you'll eventually need to start cutting up parts of this into their own functions / modules (e.g. `renderThreads`), but it's also perfectly appropriate to leave that until later -- you'll have a better idea then of what you'll **actually** need as opposed to what you **think** you'll need.

Excellent work!

File: src/index.js [hunk 4]
```diff
@@ -3,22 +3,39 @@
...
-      </li>
-    </ul>
-  `;
+import store, {getMessageSubject, getMessageSender, formatSender, formatTimestamp} from './store'
+
+function renderThreads() {
+  const { threadIds } = store.mailboxes.INBOX;
+  const { messages } = store;
+
+  if (store.mailboxes.INBOX == null) return;
```
Comment (cbothner) [inline src/index.js:12]: Returning `undefined` here without checking for undefined on line 37 means that you’ll get the literal string “undefined” in your sidebar if there are no threads. I’d return `''`

File: src/index.js [hunk 5]
```diff
@@ -3,22 +3,39 @@
...
+
+  return threadIds.map( ind => {
+    const subject = getMessageSubject(messages[ind]);
+    const snippet = messages[ind].snippet;
+    const sender = getMessageSender(messages[ind]);
+    const timestamp = formatTimestamp(messages[ind].internalDate);
```
Comment (cbothner) [inline src/index.js:18]: Teeny little thing but I’d DRY this up by making a temporary variable for `messages[ind]`

File: src/store/index.js [hunk 6]
```diff
@@ -22,4 +22,18 @@ export function getMessageSubject(message: Message): ?string {
   return subjectHeader && subjectHeader.value;
 }
 
+export function getMessageSender(message: Message): ?string {
+  const { headers } = message.payload;
+  const senderHeader = headers.find(header => header.name === 'From');
+  return senderHeader && senderHeader.value;
+}
+
+export function formatSender(sender: String){
```
Comment (cbothner) [inline src/store/index.js:31]: Flow uses lowercase `string`, for one. But I think you’re passing in a maybe-string (`?string`) too.

File: src/index.js [hunk 7]
```diff
@@ -3,22 +3,39 @@
...
+
+    return `<li>
+    <button class="email-item" type="button">
+        <div class="sender-details">
+          <p class="email">${formatSender(sender)}</p>
+          <span class="time">${timestamp}</span>
+        </div>
+        <p class="email-subject">${subject}</p>
```
Comment (cbothner) [inline src/index.js:26]: What would you do if subject is null?

**Review by cbothner — APPROVED** [review]
This is a phenomenal PR. Very well factored, idiomatic JavaScript. Readable, clear, focused. Good job. 

These few comments seem mostly to stem from “correctness” problems (that is, technicalities that would arise from exceptions to the norm that you might see in real data.) I think flow should have several many of these out. It may be annoying sometimes but it’s useful 😂

But yeah, very impressive solution!

**Review by pdhoopr — APPROVED** [review]
This is really good! I think others have already covered the points I would have touched on.

The only possible thing I can think of is that at some point, you might consider separating `format` functions out of your `store`, which is usually reserved for just working with the data (get/set, manipulation, etc.). This might just be my preference/interpretation! We can touch on it at another point, definitely not important right now.

**Pull request merged**